src/com/ms/silverking/cloud/meta/WatcherBase.java

```diff
@@ -21,204 +23,224 @@
+    lastCheckMillis = new AtomicLong();
+    timerTask = new SafeTimerTask(new WatcherTimerTask());
+    timer.schedule(timerTask, ThreadLocalRandom.current().nextInt(Math.max(minSleepMS, (int) maxInitialSleep + 1)),
+        intervalMillis);
+  }
+
+  public WatcherBase(MetaClientCore metaClientCore, String basePath, long intervalMillis, long maxInitialSleep) {
+    this(metaClientCore, _timer, basePath, intervalMillis, maxInitialSleep);
+  }
+
+  public void stop() {
+    lock.lock();
+    try {
+      active = false;
+      timerTask.cancel();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public boolean isActive() {
+    return active;
+  }
+
+  protected abstract void _doCheck() throws KeeperException;
+
+  protected void doCheck(long curTimeMillis) throws KeeperException {
+    if (active) {
+      lastCheckMillis.set(curTimeMillis);
+      _doCheck();
     }
-    
-    public void stop() {
-        lock.lock();
-        try {
-            active = false;
-            timerTask.cancel();
-        } finally {
-            lock.unlock();
-        }
+  }
+
+  protected void doCheck() throws KeeperException {
+    if (active) {
+      doCheck(SystemTimeUtil.skSystemTimeSource.absTimeMillis());
     }
-    
-    public boolean isActive() {
-        return active;
+  }
+
+  public void timerRang() throws KeeperException {
+    long curTimeMillis;
+
+    curTimeMillis = SystemTimeUtil.skSystemTimeSource.absTimeMillis();
+    if (curTimeMillis - lastCheckMillis.get() < minIntervalMillis) {
+      Log.fine("Ignoring doCheck() as last call was < intervalMillis");
+    } else {
+      doCheck(curTimeMillis);
     }
-    
-    protected abstract void _doCheck() throws KeeperException;
-    
-    protected void doCheck(long curTimeMillis) throws KeeperException {
-        if (active) {
-            lastCheckMillis.set(curTimeMillis);
-            _doCheck();
-        }
+  }
+
+  class WatcherTimerTask extends TimerTask {
+    WatcherTimerTask() {
     }
-    
-    protected void doCheck() throws KeeperException {
+
+    @Override
+    public void run() {
+      lock.lock();
+      try {
         if (active) {
-            doCheck(SystemTimeUtil.systemTimeSource.absTimeMillis());
+          timerRang();
         }
+      } catch (Exception e) {
+        Log.logErrorWarning(e);
+      } finally {
+        lock.unlock();
+      }
     }
-    
-    public void timerRang() throws KeeperException {
-        long    curTimeMillis;
-        
-        curTimeMillis = SystemTimeUtil.systemTimeSource.absTimeMillis();
-        if (curTimeMillis - lastCheckMillis.get() < minIntervalMillis) {
-            Log.fine("Ignoring doCheck() as last call was < intervalMillis");
-        } else {
-            doCheck(curTimeMillis);
-        }
+  }
+
+  public void process(WatchedEvent event) {
+    if (active) {
+      try {
+        watchedEventQueue.put(new EventAndWatcher(event, this));
+      } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)